File: 백준/Silver/3085. 사탕 게임/사탕 게임.py
# N개의 줄에 사탕의 색상이 주어진다
# 빨간색 C, 파란색 P, 초록색 Z, 노란색 Y

import sys

result = 0

def check_graph(graph):
    
    global result
    
    temp = graph[0][0]
    count = 0
    
    for i in range(N):
        count = 0
        temp = graph[i][0]
        for j in range(N):
            if graph[i][j] == temp:
                count += 1
            else:
                temp = graph[i][j]
                count = 1
            result = max(result, count)
                
        count = 0
        temp = graph[0][i]
        for j in range(N):
            if graph[j][i] == temp:
                count += 1
            else:
                temp = graph[j][i]
                count = 1
            result = max(result, count)

N = int(sys.stdin.readline())

graph = []
for i in range(N):
    graph.append(list(sys.stdin.readline().strip()))

# 우 하
dx = [1, 0]
dy = [0, 1]

for i in range(N):
    for j in range(N):
        for k in range(2):
            ndx = dx[k] + j
            ndy = dy[k] + i
            if 0 <= ndx < N and 0 <= ndy < N:
                temp = graph[ndy][ndx]
                graph[ndy][ndx] = graph[i][j]
                graph[i][j] = temp
                check_graph(graph)
                temp = graph[ndy][ndx]
                graph[ndy][ndx] = graph[i][j]
                graph[i][j] = temp
                
                
print(result)

File: 백준/Silver/3085. 사탕 게임/사탕 게임.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    static int N;
    static int max = Integer.MIN_VALUE;
    static char[][] graph;

    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        String s = bf.readLine();
        N = Integer.parseInt(s);
        graph = new char[N][N];


        for(int i=0; i<N; i++){
            s = bf.readLine();
            for(int j=0; j<N; j++){
                graph[i][j] = s.charAt(j);
            }
        }


        // 가로 위치가 다르면 사탕 바꿈
        for(int i=0; i<N; i++){
            for(int j=0; j<N-1; j++){
                if(graph[i][j] != graph[i][j+1]){
                    swapCandy(i, j, i, j+1);
                    changeCnady(i, j, i, j+1);
                    swapCandy(i, j, i, j+1);
                }
            }
        }

        // 세로 위치가 다르면 사탕 바꿈
        // 가로 위치가 다르면 사탕 바꿈
        for(int j=0; j<N; j++){
            for(int i=0; i<N-1; i++){
                if(graph[i][j] != graph[i+1][j]) {
                    swapCandy(i, j, i+1, j);
                    changeCnady(i, j, i+1, j);
                    swapCandy(i, j, i+1, j);
                }
            }
        }

        if(max == Integer.MIN_VALUE) System.out.println(N);
        else System.out.println(max);

    }

    // 바꾸고 계산하는 메서드
    public static void changeCnady(int y1, int x1, int y2, int x2){
        // 가로 확인
        for(int i=0; i<N; i++){
            int count = 1;
            for(int j=0; j<N-1; j++){
                if(graph[i][j] == graph[i][j+1]) count++;
                else {
                    max = Math.max(max, count);
                    count = 1;
                }
            }
            max = Math.max(max, count);
        }

        // 세로 확인
        for(int j=0; j<N; j++){
            int count = 1;
            for(int i=0; i<N-1; i++){
                if(graph[i][j] == graph[i+1][j]) count++;
                else {
                    max = Math.max(max, count);
                    count = 1;
                }
            }
            max = Math.max(max, count);
        }
    }

    public static void backUp(char[][] back){
        for(int i=0; i<N; i++){
            for(int j=0; j<N; j++){
                graph[i][j] = back[i][j];
            }
        }
    }

    public static void swapCandy(int y1, int x1, int y2, int x2){
        char temp = graph[y1][x1];
        graph[y1][x1] = graph[y2][x2];
        graph[y2][x2] = temp;
    }

}
